ext/random/random.c

```diff
@@ -88,19 +88,16 @@ static inline uint32_t rand_range32(const php_random_algo *algo, php_random_stat
 	size_t total_size = 0;
 	uint32_t count = 0;
 
-	result = algo->generate(status);
-	total_size = status->last_generated_size;
-	if (status->last_unsafe) {
-		return 0;
-	}
-	if (total_size < sizeof(uint32_t)) {
+	result = 0;
+	total_size = 0;
+	do {
 		r = algo->generate(status);
-		result = (result << status->last_generated_size) | r;
+		result = (result << (8 * status->last_generated_size)) | r;
```

Comment:
This triggers undefined behavior:

> If the value of the right operand is negative or is greater than or equal to the width of the promoted left operand, the behavior is undefined

This should either do `% 32` on the rhs of `<<` so that the shifted width doesn't exceed the integer width, or set `result` to `0` if `rhs >= 32` (that's what both gcc and clang seem to do with the UB right now).

Reply:
xref #9083